File: solutions/haskell/raindrops/1/src/Raindrops.hs
module Raindrops (convert) where

import Data.Foldable (fold)

convert :: Int -> String
convert n = case fold [go 3, go 5, go 7] of
  Just x -> x
  Nothing -> show n
  where
    go r = case n `rem` r of
      0 -> lookup r [(3, "Pling"), (5, "Plang"), (7, "Plong")]
      _ -> Nothing

File: solutions/cpp/raindrops/1/raindrops.cpp
#include "raindrops.h"

#include <sstream>

namespace raindrops {

    std::string convert(int n) {
        const bool by_3 = n % 3 == 0;
        const bool by_5 = n % 5 == 0;
        const bool by_7 = n % 7 == 0;

        std::stringstream ss;
        if (by_3) ss << "Pling";
        if (by_5) ss << "Plang";
        if (by_7) ss << "Plong";
        if (!by_3 && !by_5 && !by_7) ss << n;

        return ss.str();
    }

}  // namespace raindrops
